Source/Control/adsr.h

```diff
@@ -31,31 +31,45 @@ enum
 Original author(s) : Paul Batchelor
 
 Ported from Soundpipe by Ben Sergentanis, May 2020
+ 
+Remake by Steffan DIedrichsen, May 2021
 */
 class Adsr
 {
   public:
-    Adsr() {}
+    Adsr () {}
     ~Adsr() {}
     /** Initializes the Adsr module.
         \param sample_rate - The sample rate of the audio engine being run. 
     */
-    void Init(float sample_rate);
+    void Init(float sample_rate, int blockSize = 1);
 
+    /**
+     \function Retrigger forces the envelope back to attack phase
+     \param hard  resets the hosrory to zero, results in a click.
+     */
+    
+    void  Retrigger(bool hard);
 
+    
     /** Processes one sample through the filter and returns one sample.
         \param gate - trigger the envelope, hold it to sustain 
     */
+    
     float Process(bool gate);
 
     /** Sets time
         Set time per segment in seconds
     */
-    inline void SetTime(int seg, float time) { seg_time_[seg] = time; }
+    void SetTime(int seg, float time);
```

Comment:
I'd love to change this API while we're at it. Right now, someone could pass all sorts of nonesense for the `seg` argument, e.g. -1000 or `ADSR_SEG_SUSTAIN` and create hard faults easily. I see two ways how this could be improved.
1. Provide individual functions for attack decay and sustain (+ maybe another function that sets them all at once). This has the benefit that we can add additional arguments (like segment shapes, etc.) later while still maintaining the same API if we provide default values for the new arguments.
2. Use an `enum class` for the `seg` argument.

Reply:
I'm a bit reluctant to change the API. We could extend it and issue a deprecation warning for the existing one. Otherwise we have to maintain a few examples, which use the adsr.

Reply:
Hmm, we could mark the existing function as deprecated and add the new functions that also have the shape parameters. Then we could redirect the now deprecated `SetTime` to use the new functions internally.

Reply:
I introduced new methods, but kept the existing one to avoid API changes.